File: sdk/csharp/src/SAML.cs
using System;
using System.Net.Http;
using System.Xml.Linq;

namespace GGID
{
    /// <summary>
    /// GGID SAML SDK for C#.
    /// Generates SP metadata for IdP integration.
    /// </summary>
    public class SAML
    {
        public string EntityId { get; }
        public string AcsUrl { get; }
        public string SloUrl { get; }

        /// <param name="entityId">SP Entity ID (e.g. https://myapp.example.com/saml)</param>
        /// <param name="acsUrl">Assertion Consumer Service URL</param>
        /// <param name="sloUrl">Single Logout URL (optional)</param>
        public SAML(string entityId, string acsUrl, string sloUrl = "")
        {
            if (string.IsNullOrEmpty(entityId))
                throw new ArgumentException("entityId is required");
            if (string.IsNullOrEmpty(acsUrl))
                throw new ArgumentException("acsUrl is required");

            EntityId = entityId;
            AcsUrl = acsUrl;
            SloUrl = sloUrl;
        }

        /// <summary>
        /// Generate SP metadata XML string.
        /// </summary>
        public string GenerateSPMetadata()
        {
            var ns = XNamespace.Get("urn:oasis:names:tc:SAML:2.0:metadata");
            var validUntil = DateTime.UtcNow.AddYears(1).ToString("o");

            var doc = new XDocument(
                new XElement(ns + "EntityDescriptor",
                    new XAttribute("entityID", EntityId),
                    new XAttribute("validUntil", validUntil),
                    new XElement(ns + "SPSSODescriptor",
                        new XAttribute("protocolSupportEnumeration", "urn:oasis:names:tc:SAML:2.0:protocol"),
                        new XElement(ns + "NameIDFormat", "urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress"),
                        new XElement(ns + "AssertionConsumerService",
                            new XAttribute("Binding", "urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST"),
                            new XAttribute("Location", AcsUrl),
                            new XAttribute("index", 0)
                        )
                    )
                )
            );

            // Add SLO if configured
            if (!string.IsNullOrEmpty(SloUrl))
            {
                var spDescriptor = doc.Root.Element(ns + "SPSSODescriptor");
                spDescriptor.Add(new XElement(ns + "SingleLogoutService",
                    new XAttribute("Binding", "urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect"),
                    new XAttribute("Location", SloUrl)
                ));
            }

            return doc.Declaration + "\n" + doc.ToString();
        }

        /// <summary>
        /// Fetch IdP metadata from GGID instance.
        /// </summary>
        /// <param name="ggidBaseUrl">e.g. https://ggid.example.com</param>
        public static string FetchIdPMetadata(string ggidBaseUrl)
        {
            var url = ggidBaseUrl.TrimEnd('/') + "/saml/idp/metadata";
            using var client = new HttpClient();
            return client.GetStringAsync(url).GetAwaiter().GetResult();
        }
    }
}

File: sdk/react/src/saml.ts
/**
 * SAML SP utilities for GGID React SDK
 *
 * Generate SP metadata, fetch IdP metadata, and build SAML auth request URLs.
 */

export interface SAMLConfig {
  entityId: string;
  acsUrl: string;
  sloUrl?: string;
  signRequests?: boolean;
}

export function generateSPMetadata(config: SAMLConfig): string {
  const slo = config.sloUrl
    ? `  <SingleLogoutService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect" Location="${escapeXml(config.sloUrl)}" />\n`
    : "";
  return `<?xml version="1.0" encoding="UTF-8"?>
<EntityDescriptor xmlns="urn:oasis:names:tc:SAML:2.0:metadata" entityID="${escapeXml(config.entityId)}">
  <SPSSODescriptor protocolSupportEnumeration="urn:oasis:names:tc:SAML:2.0:protocol">
    <NameIDFormat>urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress</NameIDFormat>
    <AssertionConsumerService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST"
      Location="${escapeXml(config.acsUrl)}" index="0" />
${slo}  </SPSSODescriptor>
</EntityDescriptor>`;
}

export async function fetchIdPMetadata(ggidBaseUrl: string): Promise<string> {
  const res = await fetch(`${ggidBaseUrl.replace(/\/$/, "")}/saml/metadata`);
  if (!res.ok) throw new Error(`Failed to fetch IdP metadata: ${res.status}`);
  return res.text();
}

export function parseEntityId(metadataXml: string): string | null {
  const match = metadataXml.match(/entityID="([^"]+)"/);
  return match ? match[1] : null;
}

export function parseSsoUrl(metadataXml: string): string | null {
  const match = metadataXml.match(/SingleSignOnService[^>]*Location="([^"]+)"/);
  return match ? match[1] : null;
}

export function buildAuthnRequestUrl(
  ssoUrl: string, entityId: string, acsUrl: string, relayState?: string,
): string {
  const id = `_${Math.random().toString(36).slice(2, 12)}${Date.now().toString(36)}`;
  const request = `<samlp:AuthnRequest xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol" ID="${id}" Version="2.0" IssueInstant="${new Date().toISOString()}" Destination="${escapeXml(ssoUrl)}" AssertionConsumerServiceURL="${escapeXml(acsUrl)}"><saml:Issuer xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion">${escapeXml(entityId)}</saml:Issuer></samlp:AuthnRequest>`;
  const encoded = typeof Buffer !== "undefined"
    ? Buffer.from(request).toString("base64")
    : btoa(request);
  const params = new URLSearchParams({ SAMLRequest: encoded });
  if (relayState) params.set("RelayState", relayState);
  const sep = ssoUrl.includes("?") ? "&" : "?";
  return `${ssoUrl}${sep}${params.toString()}`;
}

function escapeXml(str: string): string {
  return str.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;").replace(/'/g, "&apos;");
}
